### PR TITLE
RestChannel#publish: implement params (RSL1l)

### PR DESCRIPTION
**Pull request by SimonWoolf**
Fixes https://github.com/ably/ably-ruby/issues/209

### CODE REVIEW

File: lib/ably/rest/channel.rb [hunk 1]
```diff
@@ -65,17 +64,25 @@ def initialize(client, name, channel_options = {})
...
       #
-      def publish(name, data = nil, attributes = {})
-        messages = if name.kind_of?(Enumerable)
-          name
+      #   # Publish a single Ably::Models::Message object, with a query params
+      #   # specifying quickAck: true
+      #   message = Ably::Models::Message(name: 'click', { x: 1, y: 2 })
+      #   channel.publish message, {quickAck: 'true'}
+      #
+      def publish(first, second = nil, third = {})
```
Comment (TheSmartnik) [inline lib/ably/rest/channel.rb:72]: `name, data, attributes` made a lot more sense than first, second, third, to be honest. Can we keep those?

File: lib/ably/rest/channel.rb [hunk 2]
```diff
@@ -103,7 +110,11 @@ def publish(name, data = nil, attributes = {})
           end
         end
 
-        response = client.post("#{base_path}/publish", payload.length == 1 ? payload.first : payload)
+        options = {}
+        if qs_params
+          options[:qs_params] = qs_params
```
Comment (TheSmartnik) [inline lib/ably/rest/channel.rb:115]: Oneliner like would be more readable, wdyt?
`options = { qs_params: qa_params } if qs_params`

File: lib/ably/rest/channel.rb (hunk 1 above)
Comment (SimonWoolf) [reply lib/ably/rest/channel.rb:72]: My reasoning was they make more sense for one out of the three overloaded methods, but are completely wrong for the other two, wheras deliberately-meaningless names act as a reminder that the parameters can be multiple things. WDYT?

File: lib/ably/rest/channel.rb (hunk 2 above)
Comment (SimonWoolf) [reply lib/ably/rest/channel.rb:115]: sure

File: lib/ably/rest/channel.rb (hunk 1 above)
Comment (TheSmartnik) [reply lib/ably/rest/channel.rb:72]: @SimonWoolf I understand that `first` can be multiple things, but what about `second`/`third`. It's still data and attributes, is it not?

Comment (SimonWoolf) [reply lib/ably/rest/channel.rb:72]: no - second/third are only data/attributes respectively if using the `publish(name, data)` form. If using the `publish(Message | Message[], params)` form then `second` is query params, not data payload, and `third` is ignored

Comment (TheSmartnik) [reply lib/ably/rest/channel.rb:72]: Cool then, thanks!

**Review by TheSmartnik — APPROVED** [review]

File: lib/ably/rest/channel.rb [hunk 3]
```diff
@@ -65,17 +64,25 @@ def initialize(client, name, channel_options = {})
       #   ]
       #   channel.publish messages
       #
-      def publish(name, data = nil, attributes = {})
-        messages = if name.kind_of?(Enumerable)
-          name
+      #   # Publish a single Ably::Models::Message object, with a query params
+      #   # specifying quickAck: true
+      #   message = Ably::Models::Message(name: 'click', { x: 1, y: 2 })
+      #   channel.publish message, {quickAck: 'true'}
```
Comment (mattheworiordan) [inline lib/ably/rest/channel.rb:70]: Pedantic, but `channel.publish message, quickAck: 'true'` is more Ruby-esque
If you must have `{}` then this lib has stuck with spaces i.e. `channel.publish message, { quickAck: 'true' }`

File: lib/ably/rest/channel.rb [hunk 4]
```diff
@@ -65,17 +64,25 @@ def initialize(client, name, channel_options = {})
...
-
-          name = ensure_utf_8(:name, name, allow_nil: true)
-          ensure_supported_payload data
-          [{ name: name, data: data }.merge(attributes)]
+          # (name, data, attributes) form
+          first = ensure_utf_8(:name, first, allow_nil: true)
+          ensure_supported_payload second
+          # RSL1h - attributes as an extra method parameter is extra-spec but need to
+          # keep it for backcompat until version 2
+          [[{ name: first, data: second}.merge(third)], nil]
```
Comment (mattheworiordan) [inline lib/ably/rest/channel.rb:85]: Missing space here too after `second

File: lib/ably/rest/channel.rb [hunk 5]
```diff
@@ -40,15 +40,14 @@ def initialize(client, name, channel_options = {})
...
-      # @param data [String, ByteArray, nil]   The message payload unless an Array of [Ably::Model::Message] objects passed in the first argument
-      # @param attributes [Hash, nil]   Optional additional message attributes such as :extras, :id, :client_id or :connection_id, applied when name attribute is nil or a string
+      # Publish one or more messages to the channel. Three overloaded forms
+      # @param name [String, Array<Ably::Models::Message|Hash>, Ably::Models::Message, nil]   The event name of the message to publish, or an Array of [Ably::Model::Message] objects or [Hash] objects with +:name+ and +:data+ pairs, or a single Ably::Model::Message object
+      # @param data [String, ByteArray, Hash, nil]   The message payload unless an Array of [Ably::Model::Message] objects passed in the first argument, in which case an optional hash of query parameters
+      # @param attributes [Hash, nil]   Optional additional message attributes such as :extras, :id, :client_id or :connection_id, applied when name attribute is nil or a string (Deprecated, will be removed in 1.2 in favour of constructing a Message object)
```
Comment (mattheworiordan) [inline lib/ably/rest/channel.rb:46]: Why would this be removed in 1.2?  That's a breaking change and unnecessary.

**Review by mattheworiordan — CHANGES_REQUESTED** [review]
Few minor comments.
Interested in the decision to sneak in a deprecation and statement about removing a feature in 1.2, when this has been used widely by customers, is idiomatic for Ruby, and is a breaking change.

File: lib/ably/rest/channel.rb (hunk 5 above)
Comment (SimonWoolf) [reply lib/ably/rest/channel.rb:46]: @mattheworiordan Sorry, that's my mistake - it should say 2.0, as it's a breaking change. I'll fix that.

The spec item that requires it be deprecated is https://docs.ably.io/client-lib-development-guide/features/#RSL1h . The docs pr for that was https://github.com/ably/docs/pull/742 , and the issues it addressed were https://github.com/ably/docs/issues/644 and https://github.com/ably/docs/issues/737#issuecomment-538471207 , those issues have all the discussion that lead to us agreeing to deprecate messageAttributes as a method parameter of publish().

**Comment by SimonWoolf** [comment]
>  Interested in the decision to sneak in a deprecation and statement about removing a feature in 1.2, when this has been used widely by customers, is idiomatic for Ruby, and is a breaking change.

Discussion on this in thread https://github.com/ably/ably-ruby/pull/210#discussion_r543497287

**Comment by SimonWoolf** [comment]
Running tests locally failed with some presence untilAttach tests failing (expectedly; they need to be removed), plus `Ably::Realtime::Channel Message using JSON protocol encoding and decoding encrypted messages publishing on an unencrypted channel and subscribing on an encrypted channel with another client does not attempt to decrypt the message` failing. Not a regression from this PR, so I'm going to leave those for @TheSmartnik to look into as part of getting CI working again wrt travis issues.

<details>

```
Failures:

  1) Ably::Realtime::Channel Message using JSON protocol encoding and decoding encrypted messages publishing on an unencrypted channel and subscribing on an encrypted channel with another client does not attempt to decrypt the message
     Failure/Error: Unable to find matching line from backtrace
     Timeout::Error:
       execution expired
     # ./spec/support/event_machine_helper.rb:15:in `block in run_reactor'
     # ./spec/support/event_machine_helper.rb:14:in `run_reactor'
     # ./spec/support/event_machine_helper.rb:128:in `block (3 levels) in <top (required)>'
     # ./spec/support/rest_testapp_before_retry.rb:6:in `block (2 levels) in <top (required)>'
     # ./spec/support/event_machine_helper.rb:135:in `block (2 levels) in <top (required)>'

  2) Ably::Realtime::Presence history using JSON protocol with option until_attach: true retrieves all presence messages before channel was attached
     Failure/Error: Unable to find matching line from backtrace
     Timeout::Error:
       execution expired
     # ./spec/support/event_machine_helper.rb:15:in `block in run_reactor'
     # ./spec/support/event_machine_helper.rb:14:in `run_reactor'
     # ./spec/support/event_machine_helper.rb:128:in `block (3 levels) in <top (required)>'
     # ./spec/support/rest_testapp_before_retry.rb:6:in `block (2 levels) in <top (required)>'
     # ./spec/support/event_machine_helper.rb:135:in `block (2 levels) in <top (required)>'

  3) Ably::Realtime::Presence history using JSON protocol with option until_attach: true and two pages of messages retrieves two pages of messages before channel was attached
     Failure/Error: Unable to find matching line from backtrace
     Timeout::Error:
       execution expired
     # ./spec/support/event_machine_helper.rb:15:in `block in run_reactor'
     # ./spec/support/event_machine_helper.rb:14:in `run_reactor'
     # ./spec/support/event_machine_helper.rb:128:in `block (3 levels) in <top (required)>'
     # ./spec/support/rest_testapp_before_retry.rb:6:in `block (2 levels) in <top (required)>'
     # ./spec/support/event_machine_helper.rb:135:in `block (2 levels) in <top (required)>'

  4) Ably::Realtime::Presence history using MsgPack protocol with option until_attach: true retrieves all presence messages before channel was attached
     Failure/Error: Unable to find matching line from backtrace
     Timeout::Error:
       execution expired
     # ./spec/support/event_machine_helper.rb:15:in `block in run_reactor'
     # ./spec/support/event_machine_helper.rb:14:in `run_reactor'
     # ./spec/support/event_machine_helper.rb:128:in `block (3 levels) in <top (required)>'
     # ./spec/support/rest_testapp_before_retry.rb:6:in `block (2 levels) in <top (required)>'
     # ./spec/support/event_machine_helper.rb:135:in `block (2 levels) in <top (required)>'

  5) Ably::Realtime::Presence history using MsgPack protocol with option until_attach: true and two pages of messages retrieves two pages of messages before channel was attached
     Failure/Error: Unable to find matching line from backtrace
     Timeout::Error:
       execution expired
     # ./spec/support/event_machine_helper.rb:15:in `block in run_reactor'
     # ./spec/support/event_machine_helper.rb:14:in `run_reactor'
     # ./spec/support/event_machine_helper.rb:128:in `block (3 levels) in <top (required)>'
     # ./spec/support/rest_testapp_before_retry.rb:6:in `block (2 levels) in <top (required)>'
     # ./spec/support/event_machine_helper.rb:135:in `block (2 levels) in <top (required)>'

Finished in 48 minutes 10 seconds (files took 3.79 seconds to load)
3642 examples, 5 failures, 9 pending
```

</details>

**Pull request merged**